File: Stack/032.Longest-Valid-Parentheses/032.Longest-Valid-Parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        Stack<Integer> stack = new Stack<>();
        int ret = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else if (!stack.isEmpty() && s.charAt(stack.peek()) == '(') {
                stack.pop();
                ret = Math.max(ret, stack.isEmpty() ? i + 1 : i - stack.peek());
            } else {
                stack.push(i);
            }
        }
        return ret;
    }
}

File: Stack/032.Longest-Valid-Parentheses/032.Longest-Valid-Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        result = 0
        stack = []
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            elif len(stack) > 0 and s[stack[-1]] == '(':
                stack.pop()
                result = max(result, i + 1 if len(stack) == 0 else i - stack[-1])
            else:
                stack.append(i)
        return result
